### ISSUE TITLE
Ad Insertion

### ISSUE THREAD

**Issue by fluffy** [moq-wg/moq-transport#527]
How does insertion of advertisements work and do we need any changes in MoQT to support that ?

**Comment by fluffy** [comment]
@wilaw thoughts ?

**Comment by kixelated** [comment]
I don't have a good answer for this.

HLS makes it "easy" to perform ad insertion because there's a decoupling of the manifest and edge server. You can run a custom manifest that, on a per segment basis, instructs users which segment to download next. This doesn't require any cooperation with the CDN that actually serves the segments.

However due to the subscription-based nature of MoQ... it's more complicated. The only idea I have is to create a new track at possible ad insertion points and force the viewer to resubscribe, consulting the ad server for the name of the new track. It's kind of gross but it would work.

**Comment by wilaw** [comment]
I have though about this a bunch because support for advertising is crucial for the commercial viability of MOQT-based streaming formats. I had a side meeting with Yospace CTO David Springall  a few months back to talk through some conceptual ideas. He provided this diagram of how MOQ might integrate into a low latency ad insertion workflow:

![PastedGraphic-1](https://github.com/user-attachments/assets/ee875ea7-8e0c-4723-afb9-c2c9726caa62)

The OTT ad industry is moving towards server-guided ad insertion, a mode of insertion which the client performs the insert but using instructions provided by a server. 

There are some key points in all of this:
1. We need the concept of a playback session. Luckily, we have a good proxy for this with our Wt or QUIC session establishment. We assign this session a random ID and it registers itself, either via HTTP or MOQT (I am in favor of the latter) with an ad backend. This backend will understand the previously registered broadcast and then be able allocate ad placement opportunties to each individual session. 
2. Interstitial ads are VOD content, even when placed into live streams. This means we can FETCH them. 
3. MOQT encoders will need to be SCTE compliant, as those markers are the dominant placement signaling mechanism in use today.  In my draft design of a timeline track, I [showed](https://docs.google.com/presentation/d/1MgcASdH_b8gUpMHjO-05Gbk_0CP8yw0RQO9gVO0UHrg/edit#slide=id.g21bc8d0f159_0_23) how an off-the-shelf SCTE-35 marker could be embedded. 
4. It is important that we are compliant with existing ad inventory. This is almost universally stored as ISOBMFF or TS files. It would be possible to create a proxy gateway to these ad systems which vend MOQT track variants of these assets, including trans-packaging to LOC or other container formats. 
5. We can use catalogs to instruct clients on when to insert ads. In order to provide user-personalized advertising, we need to be able to generate unique catalogs for each user. This must be done at the edge. To enable this, we should introduce the concept of an application layer at the outermost relay server. This app layer will provide functionality similar to the serverless edge compute offered with HTTP today. It will be able to subscribe to a base catalog and then modify it for each connection. It should also be able to modify Objects themselves. We can use this for many customization features (per-session watermarking, codec/format restrictions based on device type, content offering suggestions etc). 

MOQT offers some benefits for advertising that we don't have in the HTTP workflow:
1.  Having millions of client report ad playback quartiles during synchronized ad placements means that the ad reporting services need to be highly scalable. Today this is done with separate data collection networks. The directional nature of MOQT connections means that client-side ad reporting can be made though the same distribution network which is serving the media. This affords much richer, reliable and easily scalable reporting. 
2. The parallel nature of MOQT means that more flexible ad displays can be implemented. Overlays are possible MOQT can also carry image assets for advertising, it does not need to be A/V. 

Yospace have signaled interest in participating in some early proofs-of-concept to start exploring these ideas.

**Comment by kota-yata** [comment]
What about SSAI? MoQT principle of not modifying stream content on the path and SSAI doesn't seem to fit together. I'm thinking something like this for SSAI-like insertion, but I don't feel like this is efficient or viable for personalization

<img width="5426" height="3042" alt="Image" src="https://github.com/user-attachments/assets/5cc45d52-9fd5-44a9-a027-e2e21080461e" />

**Comment by wilaw** [comment]
@kota-yata  - what I had in mind is explained in the diagram below. The original publisher would publish a catalog (or equivalent list of tracks). This would include the audio and video content as well as a timeline track. The timeline track describes the history of the audio and video from that publisher and its common for all subscribers, so I don't want to put any ad related data in it. For that, we add a new track which I'm calling an 'Ad insertion track". It will be filled with a list of instructions of which ads to insert and when. This ad insertion track is not coming from the original media publisher (green), but from another original publisher I'm calling the "MOQ compliant ad decisioning system" (blue). When the client subscribes to this ad track, it appends its userID to the subscription. This allows the ad decisioning system to customize an ad pod for that user. The track returns a set of insertion instructions of which ads to insert and when. This ad content is provided by a third original publisher I'm terming the "advertising content origin" (purple). 

This architecture allows the same catalog to be consumed by all end-users (good for scale) while allowing the ad insertion instructions to be customized for each end user based on their identifier which makes it compatible with current interstitial solutions (good for interop and to leverage existing ad inventory and decisioning systems). 

Let me know if you have opinions. 


![Image](https://github.com/user-attachments/assets/cb2bafa9-25c9-4c5e-b853-adde5b7ec3b6)

**Comment by kota-yata** [comment]
Thanks.
To support ad insertion at dynamic timings like the end of an inning in a baseball game, the decisioning system needs a mechanism where the ad insertion track delivers the ad content in advance, while the exact timing for insertion (i.e., timeline placement) is notified later. This way the content could be preloaded and ready, and the actual insertion point can be determined in real time based on the state of the live event or stream.

**Comment by kota-yata** [comment]
I'm thinking of the situation where the ad decisioning system is in the production studio, and there is an operator that clicks "Insert Ad" button to dynamically insert ad for certain period of time.

**Comment by gwendalsimon** [comment]
@wilaw your solution requires the client to explicitly subscribe to the ad tracks. It is not SSAI, is it?

@kota-yata in my understanding, SSAI should be implemented by an "intermediary" (see #1107), i.e. an entity that is both a Publisher and a Subscriber, but is not a Relay. 

What I have in mind is something as follow (which mixes both your approaches): When it is time for ads, the Original Publisher adds a marker in the "timeline" track. The Intermediary has subscribed to the "regular video" track and the "timeline" track. When it detects the marker in the timeline track, the Intermediary subscribes to the "MoQ compliant ad decisioning system" with the Subscriber track Id=1234. It gets the "Ad timeline" track. Then it subscribes to the Ad track #1 and subsequently Ad track #2. And finally, the intermediary publishes the Objects of both Ad tracks into the "regular video" track toward the end-subscriber.

**Comment by wilaw** [comment]
> @wilaw your solution requires the client to explicitly subscribe to the ad tracks. It is not SSAI, is it?

Indeed. It's Server Guided Ad Insertion (SGAI) which is where the OTT ad industry is moving with both Interstitials in HLS and AlternativeMPD in DASH. This has shown to be flexible and scalable. I think MOQ should both inherit this evolution (i.e not spend time with CSAI and SSAI) and be interoperable with existing SGAI ad systems and inventory.

**Comment by riedlse** [comment]
SO I still think SSAI still makes a lot of sense over SGAI at least in Paramount sense and the way our stitcher works and scales. I was c hatting with ChatGPT about having individual timelines and it was quite optimistic. I'm happy to have any of the esteemed folks on this chat tell me where it started hallucinating as I am told it can do time to time...

MoQ is not only useful when “everyone watches the same moment.” It was designed to support both live fan-out and individualized timelines, including pause, rewind, and “off-timeline” playback, without dropping out of MoQ.
Here’s the sharper breakdown:
1) Independent timelines (pause/rewind/DVR) are built into MoQ
MoQ’s core transport lets a subscriber FETCH arbitrary ranges of past media (“Standalone Fetch”), and—crucially—“Joining Fetch” lets a client backfill from the past and then flow straight back into the live subscription contiguously and without overlap. That’s exactly what you need for pause/DVR time-shift. ([IETF Datatracker](https://datatracker.ietf.org/doc/draft-ietf-moq-transport/))
Clients also choose Group Order (ascending or descending) per subscription/fetch, which is handy for “catch up” behavior (e.g., newest-first to close a gap). ([IETF Datatracker](https://datatracker.ietf.org/doc/draft-ietf-moq-transport/))
Net: With a live SUBSCRIBE + periodic JOINING FETCH (relative or absolute), you can implement pause, replay 30s, or join-midstream catch-up entirely over MoQ. No HLS/DASH manifest gymnastics required. ([IETF Datatracker](https://datatracker.ietf.org/doc/draft-ietf-moq-transport/))
2) Personalized ads do not require leaving MoQ
MoQ is publish/subscribe by track name/namespace; a server “stitcher” can publish a per-viewer track (or a shared program track + personalized ad objects) and the client just stays subscribed. The protocol makes no assumption that all viewers get identical bytes; that’s an application choice. ([IETF Datatracker](https://datatracker.ietf.org/doc/draft-ietf-moq-transport/))
If you want server-side SSAI that’s indistinguishable to the client (your goal), you simply interleave ad objects at Group boundaries on the same track. The client never switches protocols or tracks. (Operational note: like per-user HLS SSAI, personalization reduces cache dedupe; that’s an economics/architecture trade-off, not a MoQ limitation.) ([IETF Datatracker](https://datatracker.ietf.org/doc/draft-ietf-moq-transport/))
3) Where HLS/DASH are still strong (and why you may pick them)
The HLS/DASH ecosystem is mature for pure VoD, very long DVR windows, offline/downloads, and deep CDN cache leverage; tooling and players are everywhere. If you don’t need sub-second or interactive latency, they’re a safe, boring choice. ([nanocosmos](https://www.nanocosmos.net/blog/webrtc-latency/?utm_source=chatgpt.com), [Streaming Media Magazine](https://www.streamingmedia.com/Articles/Editorial/Featured-Articles/Media-Over-QUIC-and-the-Future-of-High-Quality-Low-Latency-Streaming-167165.aspx?utm_source=chatgpt.com))
But for low-latency live with time-shift and individualized experiences, MoQ gives you a single transport with built-in catch-up (Joining Fetch) and per-subscription scheduling/priority—no manifest surgery. ([IETF Datatracker](https://datatracker.ietf.org/doc/draft-ietf-moq-transport/))
4) Practical MoQ patterns for “independent viewing” you asked about
Pause/DVR: Keep a live SUBSCRIBE to video/main + audio/main. When the user seeks back 30s, issue Relative Joining Fetch for N groups behind the live edge; when done, let playback run back to the subscription’s live edge seamlessly (the draft defines this contiguity explicitly). ([IETF Datatracker](https://datatracker.ietf.org/doc/draft-ietf-moq-transport/))
Late join catch-up: Start a SUBSCRIBE in Descending group order for the last few groups to fill the buffer quickly, then switch to normal playback order. ([IETF Datatracker](https://datatracker.ietf.org/doc/draft-ietf-moq-transport/))
SSAI on one track: Stitch ads at Group (GOP) boundaries in the same track; no client switch, no protocol switch. (If you also mirror ad cues via a tiny “events” track using DATAGRAMs, beacons fire on time but the media still stays on one MoQ track.) ([IETF Datatracker](https://datatracker.ietf.org/doc/draft-ietf-moq-transport/))

Bottom line
MoQ ≠ “broadcast-only.” It supports individualized timelines (pause/rewind) with FETCH + Joining Fetch, per-subscription ordering, and deadlines/priorities. ([IETF Datatracker](https://datatracker.ietf.org/doc/draft-ietf-moq-transport/))
You don’t need to “go off MoQ” for personalized ads. Your stitcher can publish a per-viewer stitched track; the client stays on the same MoQ subscription. The trade-offs are about caching/ops, not protocol capability. ([IETF Datatracker](https://datatracker.ietf.org/doc/draft-ietf-moq-transport/))

**Comment by afrind** [comment]
It went a little off the rails in step 2:

> The protocol makes no assumption that all viewers get identical bytes; that’s an application choice

A given _Track_ MUST have identical bytes for all subscribers.  So this idea:

> a shared program track + personalized ad objects

only works if the shared program track contains insertion points and perhaps a common Location in all personalized ad Tracks (play ad 1 now).  Each subscriber could subscribe to two tracks, the common track and the personalized ad track which would have different ads.

**Comment by WendyLi0926** [comment]
Hi @wilaw , I saw a lot of diagrams in this issue, thus I'd like to confirm some MOQT naming model concepts here (which might be not that relevant though). Let me know if there's any relevant issues which are more proper to comment. Thank you very much!
Below is my understanding about the concepts of Track Namespace, Track, Group and Subgroup. I'd like to know whether my understanding is correct or not.
Some briefs are:

- Track Namespace is a live event/channel
- Track is video/audio in different languages/subtitle
- Group is the timeline division of each track
- Subgroup is used to gain some concurrency within a group while eliminating HOL problem.
- If it's about ABR algorithms, it should be the switch among Subgroups.
Do I understand these concepts correctly and thank you!

<img width="353" height="335" alt="Image" src="https://github.com/user-attachments/assets/a0cad419-63b5-4381-b41f-b5ba7e933574" />

**Comment by afrind** [comment]
> If it's about ABR algorithms, it should be the switch among Subgroups.

I think some applications think about using different Tracks for different bit-rates.  Today you can't easily SUBSCRIBE to only subgroup 0 of a Track, though with a design like #1164 you can.  Using a subgroup per "layer" can make sense, eg where you always want subgroup 0 but subgroup 1 represents an enhancement layer that is optional.

**Comment by WendyLi0926** [comment]
> > If it's about ABR algorithms, it should be the switch among Subgroups.
> 
> I think some applications think about using different Tracks for different bit-rates. Today you can't easily SUBSCRIBE to only subgroup 0 of a Track, though with a design like [#1164](https://github.com/moq-wg/moq-transport/pull/1164) you can. Using a subgroup per "layer" can make sense, eg where you always want subgroup 0 but subgroup 1 represents an enhancement layer that is optional.

Hi @afrind , thank you so much for your reply. 
I think what confuses me are really the group and subgroup concepts. My previous understanding is that a MOQ group represents GOP, however the subgroup definition [here](https://github.com/moq-wg/moq-transport/blob/fa8da30150eda7c6d6c0bf6768307959f695c947/draft-ietf-moq-transport.md?plain=1#L367) seems to me subgroup represents GOP. Could you provide some use cases or scenarios to help better understand the differences between group and subgroup?
Again, this question might be irrelevant, let me know if there's a more proper thread to raise questions, thank you very much!

**Comment by kota-yata** [comment]
Group could represent GOP in video transmission, and a subgroup could represent a single layer of SVC. At least initial discussions around subgroup were aiming for SVC usecase (e.g. [priority discussion at IETF121](https://datatracker.ietf.org/meeting/121/materials/slides-121-moq-updating-moqt-priorities-based-on-implementation-experience-00)). bit it's not like MOQ group MUST represent GOP and subgroup MUST do so; you can choose how to assign group/subgroup in a track.

**Comment by kixelated** [comment]
So the intent is:

```
track    == rendition
group    == gop
subgroup == layer
object   == frame
```

However, without the ability to subscribe to a specific subgroup at different priorities, I don't think this will actually work for SVC. Yes, the publisher can indicate a different priority per subgroup (aka SVC layer), but the subscriber can't.

My proposal is instead:
```
track    == rendition/layer
group    == gop
subgroup == 0 (or <deleted>)
object   == frame
```

The client would subscribe to each SVC layer/track independently, joining them based on the group. It makes it very easy to signal the priority of base/enhancement layers and use different settings (ex. delivery timeout).

But yeah like you said, there's no "correct" way to map media to MoQT primatives.

**Comment by afrind** [comment]
Can we keep this issue focused on Ad Insertion?  Please move the generic discussion of mapping to the object model to another issue (if something needs to change) or the list.

**Comment by kixelated** [comment]
As for ad insertion, my thought process is that you would have "content" tracks and "ad" tracks (maybe just VOD files). An ad signaling server would tell the client which namespace/track it needs to subscribe to next, including any required auth token. It's kiiiinda like rewriting the HLS playlist but a little bit trickier given the push based nature of MoQ.

At best, the client could ignore any requests to subscribe/download ads, but they wouldn't be able to get the content if the track name rotates at each interstitial. The ad signalling server would only give them access to the content track after the ad fill should have finished.

Also hi @afrind

**Comment by gwendalsimon** [comment]
> As for ad insertion, my thought process is that you would have "content" tracks and "ad" tracks (maybe just VOD files). An ad signaling server would tell the client which namespace/track it needs to subscribe to next, including any required auth token. It's kiiiinda like rewriting the HLS playlist but a little bit trickier given the push based nature of MoQ.
> 
> At best, the client could ignore any requests to subscribe/download ads, but they wouldn't be able to get the content if the track name rotates at each interstitial. The ad signalling server would only give them access to the content track after the ad fill should have finished.
> 
> Also hi [@afrind](https://github.com/afrind)

Your proposal is somewhat similar to what @wilaw described here https://github.com/moq-wg/moq-transport/issues/527#issuecomment-3127310782

It is indeed an implementation of SGAI in MOQ, which makes sense. Earlier in this conversation, some people have claimed they would like SSAI, where the "client" is not involved in the process. The only architecture I can think of is to run an "active intermediary Relay" https://github.com/moq-wg/moq-transport/issues/527#issuecomment-3131106213

**Comment by kixelated** [comment]
> > As for ad insertion, my thought process is that you would have "content" tracks and "ad" tracks (maybe just VOD files). An ad signaling server would tell the client which namespace/track it needs to subscribe to next, including any required auth token. It's kiiiinda like rewriting the HLS playlist but a little bit trickier given the push based nature of MoQ.
> > At best, the client could ignore any requests to subscribe/download ads, but they wouldn't be able to get the content if the track name rotates at each interstitial. The ad signalling server would only give them access to the content track after the ad fill should have finished.
> > Also hi [@afrind](https://github.com/afrind)
> 
> Your proposal is somewhat similar to what [@wilaw](https://github.com/wilaw) described here [#527 (comment)](https://github.com/moq-wg/moq-transport/issues/527#issuecomment-3127310782)
> 
> It is indeed an implementation of SGAI in MOQ, which makes sense. Earlier in this conversation, some people have claimed they would like SSAI, where the "client" is not involved in the process. The only architecture I can think of is to run an "active intermediary Relay" [#527 (comment)](https://github.com/moq-wg/moq-transport/issues/527#issuecomment-3131106213)

No, I'm describing something a little bit different.  I would consider it SSAI because the client has no way of getting the original broadcast, similar to SSAI via HLS playlist rewriting.

publisher produces:
```
content-abc123:  ------
ad-break:              |
content-def456:         -------------------
```

The names of these tracks and the position of the ad-break are made available to the ad signaling service via some means. They are NOT known by the client, hence the random suffix.

The ad signalling service tells all clients to `SUBSCRIBE content-abc123`. When the ad break is triggered, it tells clients:
- no fill: `SUBSCRIBE content-def456` (no content missed)
- fill: `FETCH some-ad-id`, `FETCH another-ad-id`, followed later (in real-time) by `SUBSCRIBE content-def456`.

You absolutely could pretend like the ads are live and use SUBSCRIBE instead of FETCH. I just think it quickly becomes too complicated/expensive without explicit CDN cooperation. And from personal experience, you're likely leaking something else that can be used to determine if it's advertisement (ex. VAST tracking, click URLs, "advertisement" banner, etc). 

At Twitch, we realized that we couldn't avoid leaking if an ad was active and settled for "at best, an ad blocker can show a black screen over an ad". The critical part is that `content-def456` is either unguessable or gated behind authentication only provided by the ad service.

**Comment by wilaw** [comment]
@kixelated  - I like your approach described above, however wouldn't it fall squarely in the Server-Guided bucket, since the the client is receiving real-time instructions (guidance) on which content to play? Here are my classification

- Client-side ad insertion: client is given the primary content address, ad content address and information on when to render the ads. It is trusted to perform the insertion operations. 
- Server-guided insertion: the client receives dynamic information on which content (primary and ad) to request and is still trusted to to request the correct content at the right time. 
- Server-Side ad insertion: the server performs all the work of inserting decisioning, retrieving ,and inserting the ads and the client passively plays a single stream of content. The client may be untrusted.

**Comment by gwendalsimon** [comment]
> [@kixelated](https://github.com/kixelated) - I like your approach described above, however wouldn't it fall squarely in the Server-Guided bucket, since the the client is receiving real-time instructions (guidance) on which content to play? Here are my classification
> 
> * Client-side ad insertion: client is given the primary content address, ad content address and information on when to render the ads. It is trusted to perform the insertion operations.
> * Server-guided insertion: the client receives dynamic information on which content (primary and ad) to request and is still trusted to to request the correct content at the right time.
> * Server-Side ad insertion: the server performs all the work of inserting decisioning, retrieving ,and inserting the ads and the client passively plays a single stream of content. The client may be untrusted.

Agree.
I think we should try to define the three modes in some specialized "Event timeline tracks" in WARP.

**Comment by afrind** [comment]
@wilaw - can this issue be closed now?

**Comment by wilaw** [comment]
At the last interim, we presented several presentations on advertising and debated these concepts at length.

[Advertising with MOQT ](https://docs.google.com/presentation/d/1DenP4DWcZSe42QVlrF1oyas-V8WNYVAMla2v8-1YqsQ)- Will
[Ad Insertion and Blackout Management in MOQ ](https://docs.google.com/presentation/d/128-O-Hr-jvpdsXuArlV7b-8bjP-ccZFvH7j0VRPGPNQ)- Gwendal

This [slide ](https://docs.google.com/presentation/d/1DenP4DWcZSe42QVlrF1oyas-V8WNYVAMla2v8-1YqsQ/edit?slide=id.g39f93a06033_1_300#slide=id.g39f93a06033_1_300) summarized the requirements on the core transport (which is the subject of this thread). The WG consensus was that _**there was adequate message capability within draft-16 to enabled client-side and server-guided ad insertion, as represented by various diagrams in this issue thread**_. The was no-consensus and in fact rejection of the notion of defining "application-like" behaviors for relays, such as would be required to implement a pure SSAI solution in a standard relay, in which tracks were multiplexed relay-side in a manner which was imperceptible to the client. An implementer can always build a relay which performs such an action, but it's behavior and interface would not be standardized by this transport specification. It could be implemented via a separate draft which defines an extension point. 

With these conclusions in mind, we'd like to close this issue.
